### PR TITLE
MOB-2730 [Successful adyen payment]

### PR DESCRIPTION
**Pull request by JeevanThandi**
## JIRA
[MOB-2730](https://karhoo.atlassian.net/browse/MOB-2730)

## DESCRIPTION
- 3DS 2.0 implemented 
- transaction id used to book trip

TODO
- Error handling
- Increase test coverage

## Types of changes

What types of changes does your code introduce to the project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix 
- [x] New feature 
- [ ] Tech Debt
- [ ] Breaking change 
- [ ] Documentation Update
- [ ] Other: *define what type of change it is*

## CHECKLIST

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. This is simply a reminder of what we are going to look for before merging code._
Please review the guidelines for contributing to this repository.

- [ ] All work relating to the ticket is complete (Code complete/Acceptance Criteria met)
- [ ] Lint and unit tests pass locally (if appropriate)
- [ ] Added tests that prove the fix is effective or that feature works (if appropriate)
- [ ] Documentation (if appropriate)

## SCREENSHOTS
[Add screenshots if any for changes you have made to the App/UI]

### CODE REVIEW

**Review by mohadian — DISMISSED** [review]

File: KarhooUISDK/Classes/Payment/AddCard/AdyenCardRegistrationFlow.swift [hunk 1]
```diff
@@ -112,38 +120,83 @@ extension AdyenCardRegistrationFlow: DropInComponentDelegate {
...
-            case .success(let result): self.handle(event: AdyenResponseHandler().nextStepFor(data: result.payload, transactionId: result.transactionID))
-            case .failure(let error): print("/payments error", error)
+            case .success(let result):
+                self.transactionId = result.transactionID
+                let event = self.adyenResponseHandler.nextStepFor(data: result.payload,
+                                                                  transactionId: result.transactionID)
+                self.handle(event: event)
+            case .failure(let error):
+                self.finish(result: .completed(value: .didFailWithError(error)))
+                print("/payments error", error)
```
Comment (jsm20) [inline KarhooUISDK/Classes/Payment/AddCard/AdyenCardRegistrationFlow.swift:136]: Do we need this?

File: KarhooUISDK/Classes/Payment/AddCard/AdyenCardRegistrationFlow.swift [hunk 2]
```diff
@@ -112,38 +120,83 @@ extension AdyenCardRegistrationFlow: DropInComponentDelegate {
...
         case .paymentAuthorised(let transactionId):
-            adyenDropIn?.viewController.dismiss(animated: true, completion: nil)
             let method = PaymentMethod(nonce: transactionId)
-            callback?(.completed(value: .didAddPaymentMethod(method: method)))
+            finish(result: .completed(value: .didAddPaymentMethod(method: method)))
         case .requiresAction(let action):
             adyenDropIn?.handle(action)
-        case .handleResult(let code): print("adyen haldner result: \(code)")
+        case .handleResult(let code):
+            print("adyen haldner result: \(code)")
```
Comment (jsm20) [inline KarhooUISDK/Classes/Payment/AddCard/AdyenCardRegistrationFlow.swift:199]: Also this?

File: KarhooUISDK/Classes/Payment/AddCard/PaymentFactory.swift [hunk 3]
```diff
@@ -32,4 +33,8 @@ final class PaymentFactory {
 
         return BraintreePaymentNonceProvider()
     }
+
+    func adyenEnvironment() -> Adyen.Environment {
```
Comment (jsm20) [inline KarhooUISDK/Classes/Payment/AddCard/PaymentFactory.swift:37]: Is Production going to be handled in a future PR?

File: KarhooUISDK/Classes/Payment/AddCard/AdyenCardRegistrationFlow.swift (hunk 2 above)
Comment (JeevanThandi) [reply KarhooUISDK/Classes/Payment/AddCard/AdyenCardRegistrationFlow.swift:199]: https://karhoo.atlassian.net/browse/MOB-2731 once this is in the console print should be gone!

File: KarhooUISDK/Classes/Payment/AddCard/PaymentFactory.swift (hunk 3 above)
Comment (JeevanThandi) [reply KarhooUISDK/Classes/Payment/AddCard/PaymentFactory.swift:37]: exactly, I wasn't sure as there are multiple adyen production environments (US, Australia...) I'm guessing we will need to infer user locale to pick the right prod environment... might need to have a ticket for enabling prod... ?

**Review by jsm20 — APPROVED** [review]

**Review by nurito-burrito — APPROVED** [review]

**Pull request merged**